oak_functions/loader/src/server.rs

```diff
@@ -1183,45 +1227,57 @@ mod tests {
-        uwabi_extensions: &mut Vec<Box<dyn UwabiExtension>>,
-        channel_handle: ChannelHandle,
-    ) -> Vec<u8> {
-        let endpoint = runtime_endpoint_for_channel_handle(uwabi_extensions, channel_handle);
-        endpoint.receiver.try_recv().unwrap()
-    }
-
-    // Helper function for testing to write to Endpoint associated to ChannelHandle extension in the
-    // runtime.
+    // Helper function for testing to write to the endpoint of the given UWABI Extension.
     async fn write_to_runtime_endpoint(
-        uwabi_extensions: &mut Vec<Box<dyn UwabiExtension>>,
-        channel_handle: ChannelHandle,
+        uwabi_extension: &mut Box<dyn UwabiExtension>,
         message: UwabiMessage,
     ) {
-        let endpoint = runtime_endpoint_for_channel_handle(uwabi_extensions, channel_handle);
+        let endpoint = uwabi_extension
+            .get_endpoint_mut()
+            .expect("No endpoint set for extension.");
         let result = endpoint.sender.send(message.to_vec().clone()).await;
         assert!(result.is_ok());
     }
 
-    // Helper function for testing to find the Endpoint associated to ChannelHandle in the runtime.
-    fn runtime_endpoint_for_channel_handle(
+    // Helper function for testing to read from the endpoint associated to ChannelHandle in the Wasm
+    // Module.
+    async fn read_from_wasm_module_endpoint(
+        wasm_state: &mut WasmState,
+        channel_handle: ChannelHandle,
+    ) -> UwabiMessage {
+        wasm_state
+            .channel_switchboard
+            .get_mut(&channel_handle)
+            .expect("Endpoint not set for Channel Handle")
+            .receiver
+            .recv()
+            .await
+            .expect("No message or channel closed.")
+    }
+
+    // Helper function for testing to find the Extension associated to ChannelHandle.
+    fn extension_for_channel_handle(
         uwabi_extensions: &mut Vec<Box<dyn UwabiExtension>>,
         channel_handle: ChannelHandle,
-    ) -> &mut Endpoint {
-        // Find extension associated to ChannelHandle.
-        let extension = uwabi_extensions
-            .iter_mut()
-            .find(|uwabi_extension| {
+    ) -> Box<dyn UwabiExtension> {
+        let pos_of_extension = uwabi_extensions
+            .iter()
+            .position(|uwabi_extension| {
                 let channel_handle_of_extension = uwabi_extension.get_channel_handle();
                 channel_handle_of_extension == channel_handle
             })
             .expect("No extension for channel handle.");
+        uwabi_extensions.remove(pos_of_extension)
+    }
 
-        // Get endpoint from the extension.
-        extension
-            .get_endpoint_mut()
-            .expect("No endpoint set for extension.")
+    // Helper function for testing to make an UWABI message human readable in code.
+    fn from_uwabi_message(uwabi_message: &UwabiMessage) -> String {
+        String::from_utf8(uwabi_message.clone())
```

Comment:
The UTF8 parsing / conversion seems unnecessary / misleading, given messages are just uninterpreted byte strings, but up to you if you think it adds any value.

Reply:
Ah, I see how it can be misleading. Maybe `from_uwabi_message/to_uwabi_message` should be added to the `UwabiExtension` trait? Because I do, in fact, interpret the byte as strings in the `run` method of the testing extension.

If you don't mind much, I would leave it for this PR but revisit it as soon as we get to the IDL. WDYT?

Reply:
I would not even interpret the byte as strings, that's also not what an IDL would do. Perhaps look at the first byte and switch on that value, then decide what to do with the rest of the message? Then it's obvious how that would be replaced by e.g. some code generated from the IDL, in principle.